[Core][Geometry] Add Assign and Calculate interfaces

tteschemacher opened the pull request:
yo,

welcome back with a beautiful PR after the summer break.

**Description**
This PR adds the interfaces for Set and Calculate to the geometry. They are similar as the ones from Elements and Conditions. This allows dynamic interfaces, especially with more sophisticated geometries. Like this we do not have to add new interfaces, but can access to all internal variables and compute upon them.

Normally it would be Get and Set. However, to be consistent it is Calculate and Set.

tteschemacher committed:
Add Set and Calculate functions to geometry

tteschemacher committed:
add Set and Calculate to python

tteschemacher committed:
fix return values

tteschemacher committed:
fix data types in set and calculate functions

philbucher commented on kratos/geometries/geometry.h:
```diff
@@ -634,6 +634,99 @@ class Geometry
+    /// Set with array_1d<double, 6>
+    virtual void Set(
+        const Variable<array_1d<double, 6>>& rVariable,
+        array_1d<double, 6>& Output) {}
```
```suggestion
    /// Set with array_1d<double, 3>
    virtual void Set(
        const Variable<array_1d<double, 3>>& rVariable,
        array_1d<double, 3>& Output) {}
```
I would keep the extension separate

tteschemacher replied:
To be honest, right now, I am explicitly after the array_1d<double, 2>. I'd actually like to keep it.

philbucher commented on kratos/geometries/geometry.h:
```diff
@@ -634,6 +634,99 @@ class Geometry
+    /// Set with double
+    virtual void Set(
+        const Variable<double>& rVariable,
+        double& Output) {}
```
```suggestion
        const double Input {}
```
this waythe usage is clearer

tteschemacher replied:
definitely. Thanks. I changed it

philbucher commented on kratos/geometries/geometry.h:
```diff
@@ -634,6 +634,99 @@ class Geometry
+    /// Calculate with Matrix
+    virtual void Calculate(
+        const Variable<Matrix>& rVariable,
+        Matrix& Output) {}
```
```suggestion
        Matrix& rOutput) {}
```

tteschemacher replied:
yo, changed it.

tteschemacher committed:
change variable name to Input

tteschemacher committed:
change Output to rOutput

tteschemacher reviewed (commented):
Thanks for the comments. I changed the namings, but I would prefer to keep the extended functions.

rubenzorrilla commented on kratos/python/add_geometries_to_python.cpp:
```diff
@@ -163,6 +185,24 @@ void  AddGeometriesToPython(pybind11::module& m)
     .def("Length",&GeometryType::Length)
     .def("Area",&GeometryType::Area)
     .def("Volume",&GeometryType::Volume)
+    // Calculate
```
```suggestion
    // Set
```

rubenzorrilla reviewed (commented):
I've two comments:
- Why in the `Set` methods we are passing `arrays_1d<double,X>`, `Vector` and `Matrix` by copy?
- `array_1d<double,2>`? 🙈

tteschemacher committed:
fix comment

Co-authored-by: Rubén Zorrilla <rubenzorrillamartinez@hotmail.com>

tteschemacher committed:
update copy to reference in Set methods for arrays and vectors

pooyan-dadvand reviewed (commented):
@KratosMultiphysics/technical-committee we see this addition to the interface as a way to reduce newer methods in the Geometry and we agree to their addition following these guidelines:
* Set should be renamed to SetValue to be consistent with elements and conditions
* Only the types that will be used now should be added to the interface (`double`, `array_1d<double,3>` and `Vector`) and the rest of the types should be included when we have a use case.

tteschemacher committed:
Make Calculate const

By definition SetValue can interact with the memory. However, Calculate shouldn't.

tteschemacher committed:
Changing Set to Assign

tteschemacher committed:
Change Set to Assign in python interface

rubenzorrilla reviewed (approved):
After a brief discussion in the @KratosMultiphysics/technical-committee , we agree that there are valid use cases for the `array_1d<double,2>`. However, for the moment these must stick to geometrical operations as that ones in here.
